sdks/java/io/contextualtextio/src/main/java/org/apache/beam/sdk/io/contextualtextio/ContextualTextIO.java

```diff
@@ -476,37 +478,53 @@ public void processElement(ProcessContext p) {
             if (pastRecords.containsKey(file)) {
               numRecordsBefore = pastRecords.get(file);
             }
-            p.output(KV.of(fileRange, numRecordsBefore));
+            p.output(KV.of(file, KV.of(fileRange.getValue(), numRecordsBefore)));
             pastRecords.put(file, numRecordsBefore + numRecords);
           }
         }
       }
     }
 
     static class AssignRecordNums extends DoFn<KV<KV<String, Long>, Row>, Row> {
```

Comment:
As discussed , please add more comments on AssignRecordNums in next commit.

Reply:
done